### PR TITLE
Array support

### PR DESCRIPTION
**Pull request by marcinkumorek**

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -64,8 +64,38 @@ const utils = {
...
+    splitNotation (notation) {
+        return notation.split(/\.|\b(?=\[)/);
+    },
+
+    concatNotes (notes) {
+        return notes.reduce((acc, note) => acc + (this.isArrIndex(note) ? note : `.${note}`))
+    },
+
+    getIndexNumber (notation) {
+        return +notation.replace(/[\[\]]/g, '');
```
Comment (marcinkumorek) [inline src/utils.js]: Should this be handled?
splitNotation(x)
`a["prop[]"]` 
will return
```
[
'a',
'["prop['
']"'
]
```

File: src/utils.js [hunk 2]
```diff
@@ -64,8 +64,38 @@ const utils = {
...
         return arr.length === 1
             && (arguments.length === 2 ? arr[0] === itemValue : true);
-    }
+    },
+
+    isArrIndex (note) {
+        return /^\[\d+\]$/.test(note);
+    },
+
+    splitNotation (notation) {
```
Comment (marcinkumorek) [inline src/utils.js]: Should this be handled?
`x = 'a["prop[]"]'`

`splitNotation(x)`

will return

```
[
'a',
'["prop['
']"'
]
```

File: src/utils.js [hunk 3]
```diff
@@ -64,8 +64,39 @@ const utils = {
...
+        // needs rethink
```
Comment (marcinkumorek) [inline src/utils.js]: It doesn't support nested array accessors `prop[0][1]`.
I'll try to update it tomorrow.

File: src/core/notation.js [hunk 4]
```diff
@@ -91,29 +91,45 @@ class Notation {
      *  // "car.model"  "Charger"
      *  // "car.year"  1970
      */
-    each(callback) {
-        let o = this._source,
-            keys = Object.keys(o);
+    each (callback) {
```
Comment (onury) [inline src/core/notation.js]: This yields keys such as `a[0]` instead of `[0]` which produces bugs. Pls see new tests I've attached to the PR conversation.

File: src/core/notation.js [hunk 5]
```diff
@@ -136,7 +152,7 @@ class Notation {
      *          console.log(note, levelValue); // "car.brand" "Dodge"
      *      });
      */
-    eachValue(notation, callback) {
+    eachValue (notation, callback) {
```
Comment (onury) [inline src/core/notation.js]: New `#eachValue()` tests for bracket notation fail.

**Comment by marcinkumorek** [comment]
So far I've managed to fix all tests except one: Notation.Glob .normalize().
I've hit quite a dead end because after fixing this one Notation.Glob.union() breaks.
Anyway, I'll try to fix this last test as soon as possible.

File: src/utils.js [hunk 6]
```diff
@@ -66,23 +66,55 @@ const utils = {
...
     },
 
-    isArrIndex (note) {
+    isArrIndex(note) {
         return /^\[\d+\]$/.test(note);
     },
 
-    splitNotation (notation) {
-        return notation.split(/\.|(?<=.)(?=\[(?=\d|['`"]{1}.+['"`]{1}\]))/);
+    splitNotation(notation) {
```
Comment (onury) [inline src/utils.js]: I don't think this is good. There is a better way. I'll let you know shortly.

File: src/utils.js [hunk 7]
```diff
@@ -66,23 +66,55 @@ const utils = {
...
+        return notes;
     },
 
-    concatNotes (notes) {
+    concatNotes(notes) {
         return notes.reduce((acc, note) => acc + (this.isArrIndex(note) ? note : `.${note}`))
     },
 
-    getIndexNumber (notation) {
+    getIndexNumber(notation) {
```
Comment (onury) [inline src/utils.js]: This will return `NaN` for `['a']`

File: src/core/notation.js [hunk 8]
```diff
@@ -412,7 +419,7 @@ class Notation {
      *  console.log(obj);
      *  // { notebook: "Mac", car: { brand: "Ford", model: "Mustang", year: 1970, color: "red" }, boat: "none" };
      */
-    set (notation, value, overwrite = true) {
+    set (notation, value, overwrite = true, concatArrays = false) {
```
Comment (onury) [inline src/core/notation.js]: Why are you changing the signature?

File: src/utils.js (hunk 7 above)
Comment (marcinkumorek) [reply src/utils.js]: Before this function is invoked, a notation is validated via `utils.isArrIndex` method. This way only notations that pass `/^\[\d+\]$/` regex will be passed to `getIndexNumber`.

File: src/core/notation.js (hunk 8 above)
Comment (marcinkumorek) [reply src/core/notation.js]: It was the quickest way to fix merging notations.
Here is code from tests you attached:
```
    it('Notation #merge()', () => {
        const nota = create({ a: [{ x: 1 }], b: { c: [3] } });
        let merged = nota.merge({ a: [{ y: 2 }] });
        expect(merged.value.a[0]).toEqual({ x: 1, y: 2 });
        merged = nota.merge({ b: { c: [4, 5] } });
        expect(merged.value.b.c).toEqual([3, 4, 5]);
    });
```
Without array concatenation last expect would expect value `[4, 5]`  to be equal to `[3, 4, 5]` because both `[3]` and `[4, 5]` had their `0` index taken. So second array elements would overwrite first array elements.

**Comment by marcinkumorek** [comment]
Well, you're right. Right now in case of the simple object like `{ ['weird-property']: 0 }` everything would be okay if notation would look like `a.weird-property` but if written in valid JS syntax `a['weird-property']` `utils.splitNotation` would result in `[ 'a', '[\'weird-notation\']']`.

Simply removing outter brackets from non-numeric property strings should do the trick without any major changes.

I believe `a['1']` would be recognized as object property right now. But still, brackets and quotemarks would be left intact which would break program.

**Comment by onury** [comment]
See the new branch I've been playing with for the last 2 days: 
https://github.com/onury/notation/blob/bracket-support/src/core/notation.js

Let me know what you think.

Notice how `Notation.split()` & `Notation.join()` is done now.
Also see `utils.normalizeNote()`.

These validate the notation against 1) JS variable syntax 2) proper object bracket notation 3) proper array index notation.

Removed Jasmine in favour of Jest with test-coverage. This seems crucial. Notation is a complex library, so coverage is crucial. Fixed many bugs on the way.

Couldn't look inti `Notation.Glob` yet.

**Comment by onury** [comment]
Also see this article https://mathiasbynens.be/notes/javascript-properties
Right now in `bracket-support` branch, a notation of `a[123]` will throw if `a` is not an array.

**Comment by marcinkumorek** [comment]
I've checked out `bracket-support` and tested it for a while.
I like changes to `.split()` and `.join()` functions. Also, bracket notation detection works great.
Of course some `NotationGlob` tests fails but I believe it is mostly because  of `normalize` and `toRegExp` methods.

So what is the plan now? Should I fork your `bracket-support` and help you with fixes or wait until you're done and apply changes I made on updated master?

If you want I can add comments to changes I commited to explain why specific change was needed with some example so it will be easier to understand.

**Comment by onury** [comment]
Yes @marcinkumorek. For now it's better to wait a couple more days. 
I made some more fundamental changes.

For instance, I'm not using `regexp#test()` any more to [check whether globA matches/covers globB](https://github.com/onury/notation/blob/bracket-support/src/core/notation.glob.js#L305). Since the notes (either array index notation, object bracket notation or JS var syntax) are checked strictly, regexp A will never match a wildcard in globB (in `.normalize()` for example) with regexp.

`.normalize()` is improved and passes the tests now. 
Let me fix `.union()` now, which is the only method left in `NotationGlob`.

Then we'll need the following:
- since we don't depend on `toRegExp()` anymore, check/decide whether to remove it and also `.regexp` property.
- test/revise `Notation#filter()` method
- UMD module build is broken. Fix webpack config.
- add a bit more tests maybe?

I will still mention you and your help with this in readme and changelog;  even if we don't merge branches. But if you're still willing to; I'd appreciate any help with these items. Yes you'd need to fork `bracket-support` branch. Also let me know whichever part you think, should go into `bracket-support` branch from your current, instead of the current.

A side-note: 
- Always PR for one specific thing. Otherwise it's really hard to follow (when made too many huge changes).
- Never send PR from a master. Pls name your fork e.g. `bracket-fix-webpack` or something.

cheers.

**Comment by marcinkumorek** [comment]
Thanks for the tips. As for credits, I don't really care about them and you're the one that does most of the work here ;]
So, I'll try to look into your branch in a couple days if I find some free time.
As for code from my forked repo it's hard to tell which part of it should go to your branch since your code will use different methods now and some code may be completely redundant. After I'll check your branch I'll let you know.

**Comment by onury** [comment]
Thank you. Your work's still a big help. 
Plus, discussions/ideas sometimes are better contributions than code. I really appreciate that.
Let you know when I'm done shortly.

**Comment by Kamikadze4GAME** [comment]
Hello all.
Is it live?
